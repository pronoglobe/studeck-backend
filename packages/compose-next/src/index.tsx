import * as React from 'react';

export const addInitialProps = (getInitialProps: (ctx: any) => any) => {
  // tslint:disable-next-line:variable-name
  return (Component: React.ComponentType) =>
    class extends React.Component {
      public static getInitialProps(ctx: any): any {
        return getInitialProps(ctx);
      }
      public render(): JSX.Element {
        return <Component {...this.props} />;
      }
    };
};

export type ComponentWithInitialProps = React.ComponentType<any> & {
  getInitialProps?: (ctx?: any) => any;
};

// tslint:disable-next-line:variable-name
export const composeInitialProps = (Parent: ComponentWithInitialProps) => (Child: ComponentWithInitialProps) =>
  class extends React.Component<any> {
    public static async getInitialProps(ctx?: any): Promise<{ parentProps: any; childProps: any }> {
      return {
        parentProps: Parent.getInitialProps ? await Parent.getInitialProps(ctx) : {},
        childProps: Child.getInitialProps ? await Child.getInitialProps(ctx) : {},
      };
    }
    public render(): JSX.Element {
      return (
        <Parent {...this.props} {...this.props.parentProps}>
          <Child {...this.props} {...this.props.childProps} />
        </Parent>
      );
    }
  };
