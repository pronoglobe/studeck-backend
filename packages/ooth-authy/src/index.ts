import { FullRequest, Ooth, StrategyValues } from 'ooth';
import { callbackify } from 'util';

const fetch = require('node-fetch');

const { Strategy } = require('passport-custom');

type Config = {
  name?: string;
  ooth: Ooth;
  apiKey: string;
};

export default function({ name = 'authy', ooth, apiKey }: Config): void {
  ooth.registerUniqueField(name, 'email', 'email');
  ooth.registerProfileFields(name, 'email');
  ooth.registerStrategyUniqueField(name, 'id');

  ooth.registerMethod(name, 'register', [ooth.requireNotLogged], async ({ email, cellphone, country_code }) => {
    if (!email || typeof email !== 'string') {
      throw new Error('Email is required.');
    }
    if (!cellphone || typeof cellphone !== 'string') {
      throw new Error('Invalid phone number.');
    }
    if (!country_code || typeof country_code !== 'string') {
      throw new Error('Invalid coutnry code.');
    }

    const existingUser = await ooth.getUserByUniqueField('email', email);

    if (existingUser) {
      throw new Error('This email is already registered');
    }

    const res = await fetch(`https://api.authy.com/protected/json/users/new`, {
      method: 'POST',
      body: `user[email]=${email}&user[cellphone]=${cellphone}&user[country_code]=${country_code}`,
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
        'X-Authy-API-Key': apiKey,
      },
    });
    const data = await res.json();
    if (data.errors) {
      console.error(data);
      const key = Object.keys(data.errors)[0];
      throw new Error(`${key} ${data.errors[key]}`);
    }

    await ooth.insertUser(name, {
      email,
      cellphone,
      country_code,
      id: data.user.id,
    });

    return {
      message: 'Registered successfully',
    };
  });

  ooth.registerMethod(name, 'otp', [ooth.requireNotLogged], async ({ email }) => {
    const user = await ooth.getUserByUniqueField('email', email);
    if (!user || !user[name]) {
      throw new Error('This email is not registered with this authentication strategy.');
    }
    const values = user[name] as StrategyValues;
    const res = await fetch(`https://api.authy.com/protected/json/sms/${values.id}`, {
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
        'X-Authy-API-Key': apiKey,
      },
    });
    const data = await res.json();
    if (data.errors) {
      throw new Error(data.errors[Object.keys(data.errors)[0]]);
    }

    return {
      message: 'A one-time password has been send to you via SMS.',
    };
  });

  ooth.registerPrimaryAuth(
    name,
    'login',
    [ooth.requireNotLogged],
    new Strategy(
      callbackify(async (req: FullRequest) => {
        const { email, token } = req.body;
        if (!email || typeof email !== 'string') {
          throw new Error('Invalid email.');
        }

        if (!token || (typeof token !== 'string' && typeof token !== 'number')) {
          throw new Error('Invalid token.');
        }

        const user = await ooth.getUserByUniqueField('email', email);

        if (!user || !user[name]) {
          throw new Error('This email is not registered with this authentication strategy.');
        }
        const values = user[name] as StrategyValues;

        console.log(`https://api.authy.com/protected/json/verify/${token}/${values.id}`);

        const res = await fetch(`https://api.authy.com/protected/json/verify/${token}/${values.id}`, {
          headers: {
            'Content-Type': 'application/x-www-form-urlencoded',
            'X-Authy-API-Key': apiKey,
          },
        });
        const data = await res.json();
        if (data.errors) {
          throw new Error(data.errors[Object.keys(data.errors)[0]]);
        }

        return user._id;
      }),
    ),
  );
}
