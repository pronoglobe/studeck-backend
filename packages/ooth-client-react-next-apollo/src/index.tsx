import 'isomorphic-unfetch';

import { ApolloReducerConfig, InMemoryCache } from 'apollo-cache-inmemory';
import { ApolloClient, ApolloClientOptions } from 'apollo-client';
import { ApolloLink } from 'apollo-link';
import { HttpLink } from 'apollo-link-http';
import * as React from 'react';
import { ApolloProvider, getDataFromTree } from 'react-apollo';

let client: ApolloClient<any> | null = null;

const AP = ApolloProvider as any;

const getBrowserClient = (
  uri: string,
  opts: Partial<ApolloClientOptions<any>>,
  cacheOpts?: ApolloReducerConfig,
  initialData?: any,
) => {
  const link = new HttpLink({
    uri,
    credentials: 'include',
  });
  const cache = new InMemoryCache(cacheOpts);
  if (initialData) {
    cache.restore(initialData);
  }
  return new ApolloClient({
    link,
    cache,
    ...opts,
  });
};

const getServerClient = (
  uri: string,
  cookies: { [key: string]: string },
  opts: Partial<ApolloClientOptions<any>>,
  cacheOpts?: ApolloReducerConfig,
  initialData?: any,
  linkState?: any,
) => {
  let link: ApolloLink = new HttpLink({
    uri,
  });
  if (cookies && Object.keys(cookies).length) {
    const middlewareLink = new ApolloLink((operation, forward) => {
      operation.setContext({
        headers: {
          Cookie: Object.keys(cookies)
            .map((key) => `${key}=${cookies[key]}`)
            .join('; '),
        },
      });
      return forward!(operation);
    });
    link = middlewareLink.concat(link);
  }

  const cache = new InMemoryCache(cacheOpts);
  if (initialData) {
    cache.restore(initialData);
  }

  if (linkState) {
    link = ApolloLink.from([link, linkState]);
  }

  return new ApolloClient({
    link,
    cache,
    ssrMode: true,
    ...opts,
  });
};

const getClient = (
  uri: string,
  cookies: { [key: string]: string },
  opts: Partial<ApolloClientOptions<any>>,
  cacheOpts?: ApolloReducerConfig,
  initialData?: any,
  linkState?: any,
) => {
  if (!(process as any).browser) {
    // on server, create a new client for each request
    return getServerClient(uri, cookies, opts, cacheOpts, initialData, linkState);
  }
  // on client, create singleton
  if (!client) {
    client = getBrowserClient(uri, opts, cacheOpts, initialData);
  }
  return client;
};

export type Options = {
  url: string;
  opts?: ApolloReducerConfig;
  apolloOpts?: Partial<ApolloClientOptions<any>>;
  linkState?: any;
};

type Props = { initialData?: any; childProps?: any };

export default ({ url, opts, apolloOpts, linkState }: Options) => {
  // tslint:disable-next-line variable-name
  return (Component: React.ComponentType<any> & { getInitialProps?: (ctx: any) => Promise<any> }) =>
    class extends React.Component<Props> {
      private client: ApolloClient<any>;

      constructor(props: Props) {
        super(props);
        // On server, this client won't be doing any work, because all work has been done in getInitialProps
        this.client = getClient(url, {}, apolloOpts || {}, opts, props.initialData);
      }

      public static async getInitialProps(ctx: any): Promise<{ initialData: any; childProps: any }> {
        const cookies = ctx.req && ctx.req.cookies;
        const client = getClient(url, cookies, apolloOpts, linkState || {}, opts);

        const childProps = Component.getInitialProps ? await Component.getInitialProps(ctx) : {};

        if (!(process as any).browser) {
          const url = {
            query: ctx.query,
            pathname: ctx.pathname,
            asPath: ctx.asPath,
          };
          const app = (
            <AP client={client}>
              <Component url={url} {...childProps} />
            </AP>
          );
          await getDataFromTree(app);
        }

        return {
          childProps,
          initialData: client.cache.extract(),
        };
      }

      public render(): JSX.Element {
        return (
          <AP client={this.client}>
            <Component {...this.props} {...this.props.childProps} />
          </AP>
        );
      }
    };
};

export const clear = () => (client = null);
