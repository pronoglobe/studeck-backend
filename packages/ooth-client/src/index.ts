import fetch from 'cross-fetch';
import { isEqual } from 'lodash';

const url = require('url');

declare var require: any;

export type Options = {
  url: string;
  secondaryAuthMode: 'jwt' | 'session';
  api?: ApiOptions;
  ws?: boolean;
};

export type ApiOptions = {
  url: string;
  primaryAuthMode?: 'jwt';
  secondaryAuthMode: 'jwt' | 'session';
  loginPath?: string;
  logoutPath?: string;
};

export type User = {
  _id: string;
};

export type Listener = (payload: any) => Promise<void>;
export type RequestTransformer = (request: RequestInit) => RequestInit;

export class OothClient {
  private oothUrl: string;
  private secondaryAuthMode: 'jwt' | 'session';
  private api: ApiOptions | undefined;
  private listeners: { [name: string]: Listener[] } = {};
  private user: User | undefined;
  private token: string | undefined;
  private ws: boolean;
  private started: boolean = false;

  constructor({ url, secondaryAuthMode, api, ws }: Options) {
    this.oothUrl = url;
    this.secondaryAuthMode = secondaryAuthMode;
    this.api = api;
    this.ws = !!ws;
  }

  public async start(): Promise<User | undefined> {
    if (!this.started) {
      this.started = true;
      await this.fetchUser();
      if (this.ws && typeof WebSocket !== 'undefined') {
        const urlParts = url.parse(this.oothUrl);
        const protocol = urlParts.protocol === 'https:' ? 'wss' : 'ws';
        const wsUrl = `${protocol}://${urlParts.host}${urlParts.path}/ws/user`;
        const socket = new WebSocket(wsUrl);
        socket.onerror = (err: Event) => console.error(err);
        socket.onopen = () => {};
        socket.onclose = () => {};
        socket.onmessage = ({ data }: any) => this.setUser(JSON.parse(data).user);
      }
    }

    return this.user;
  }

  public on(eventName: string, listener: Listener): void {
    if (!this.listeners[eventName]) {
      this.listeners[eventName] = [];
    }
    this.listeners[eventName].push(listener);
  }

  public unsubscribe(eventName: string, listener: Listener): void {
    if (this.listeners[eventName]) {
      this.listeners[eventName] = this.listeners[eventName].filter((l) => l !== listener);
    }
  }

  public async emit(eventName: string, payload: any): Promise<void> {
    if (this.listeners[eventName]) {
      for (const listener of this.listeners[eventName]) {
        await listener(payload);
      }
    }
  }

  public async authenticate(strategy: string, method: string, body: any): Promise<User | undefined> {
    const raw = await fetch(`${this.oothUrl}/${strategy}/${method}`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: body && JSON.stringify(body),
      credentials: 'include',
    });
    const response = await raw.json();
    if (response.status === 'error') {
      throw new Error(response.message);
    }
    const { user, token } = response;
    if (token) {
      if (this.api) {
        if (this.api.primaryAuthMode === 'jwt') {
          await fetch(`${this.api.url}${this.api.loginPath}`, {
            method: 'POST',
            headers: {
              Authorization: `JWT ${token}`,
            },
            credentials: 'include',
          });
        }
      }
      if (
        this.secondaryAuthMode === 'jwt' ||
        (this.api && (this.api.primaryAuthMode === 'jwt' || this.api.secondaryAuthMode === 'jwt'))
      ) {
        this.token = token;
      }
    }
    await this.setUser(user);

    return user;
  }

  public async method<T>(strategy: string, method: string, body?: any, headers?: any): Promise<T> {
    const actualHeaders = {
      'Content-Type': 'application/json',
    };
    if (headers) {
      Object.assign(actualHeaders, headers);
    }
    if (this.secondaryAuthMode && this.token) {
      headers.Authorization = `JWT ${this.token}`;
    }
    const raw = await fetch(`${this.oothUrl}/${strategy}/${method}`, {
      method: 'POST',
      headers: actualHeaders,
      body: body && JSON.stringify(body),
      credentials: 'include',
    });
    const response = await raw.json();
    if (response.status === 'error') {
      throw new Error(response.message);
    }
    if (response.user) {
      await this.setUser(response.user);
    }
    return response;
  }

  public async logout(): Promise<void> {
    await fetch(`${this.oothUrl}/session/logout`, {
      method: 'POST',
      credentials: 'include',
    });
    if (this.api && this.api.primaryAuthMode === 'jwt') {
      await fetch(`${this.api.url}${this.api.logoutPath}`, {
        method: 'POST',
        credentials: 'include',
      });
    }
    this.token = undefined;
    this.setUser(undefined);
  }

  public async apiCall<T>(path: string, body: any, headers: any): Promise<T> {
    if (!this.api) {
      throw new Error('No api settings.');
    }

    const actualHeaders = {
      'Content-Type': 'application/json',
    };
    if (headers) {
      Object.assign(actualHeaders, headers);
    }
    if (this.api.secondaryAuthMode === 'jwt' && this.token) {
      headers.Authorization = `JWT ${this.token}`;
    }
    const raw = await fetch(`${this.api.url}${path}`, {
      method: 'POST',
      headers: actualHeaders,
      body: body && JSON.stringify(body),
    });
    const response = await raw.json();
    if (response.status === 'error') {
      throw new Error(response.message);
    }
    return response;
  }

  public async fetchUser(headers?: any): Promise<User | undefined> {
    await this.method('user', 'user', null, headers);

    return this.user;
  }

  public async setUser(user: User | undefined): Promise<void> {
    if (!isEqual(this.user, user)) {
      this.user = user;
    }
    await this.emit('user', user);
  }
}
