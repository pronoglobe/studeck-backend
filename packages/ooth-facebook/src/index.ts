import * as FacebookStrategy from 'passport-facebook-token';
import { Ooth, StrategyValues } from 'ooth';

type Config = {
  name?: string;
  ooth: Ooth;
  clientID: string;
  clientSecret: string;
};

type Profile = {
  id: string;
  emails: {
    value: string;
  }[];
};

export default function({ name = 'facebook', ooth, clientID, clientSecret }: Config): void {
  ooth.registerUniqueField(name, 'email', 'email');
  ooth.registerProfileFields(name, 'email');
  ooth.registerStrategyUniqueField(name, 'id');
  ooth.registerPrimaryConnect(
    name,
    'login',
    [],
    new FacebookStrategy(
      {
        clientID,
        clientSecret,
      },
      (_accessToken: string, _refreshToken: string, profile: Profile, done: (e: Error | null, u: StrategyValues) => void) =>
        done(null, {
          id: profile.id,
          email: profile.emails[0].value,
        }),
    ),
  );
}
