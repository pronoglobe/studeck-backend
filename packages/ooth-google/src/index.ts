import { Ooth, StrategyValues } from 'ooth';

// tslint:disable-next-line variable-name
const GoogleStrategy = require('passport-google-id-token');

type Config = {
  name?: string;
  ooth: Ooth;
  clientID: string;
  clientSecret: string;
};

export default function({ name = 'google', ooth, clientID, clientSecret }: Config): void {
  ooth.registerUniqueField(name, 'email', 'email');
  ooth.registerProfileFields(name, 'email');
  ooth.registerPrimaryConnect(
    name,
    'login',
    [],
    new GoogleStrategy(
      {
        clientID,
        clientSecret,
      },
      (parsedToken: { payload: { email: string } }, _googleId: string, done: (e: Error | null, u: StrategyValues) => void) =>
        done(null, {
          email: parsedToken.payload.email,
        }),
    ),
  );
}
