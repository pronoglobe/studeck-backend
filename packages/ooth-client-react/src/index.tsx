import * as React from 'react';
import { getContext } from 'recompose';
import { OothClient, User } from 'ooth-client';
import * as PropTypes from 'prop-types';

type Props = {
  client: OothClient;
  initialUser: User | undefined;
};

type Status = { user: User | undefined };

export class OothProvider extends React.Component<Props, Status> {
  public static childContextTypes: any = {
    oothClient: PropTypes.object.isRequired,
    user: PropTypes.object,
  };

  constructor(props: Props) {
    super(props);
    this.state = {
      user: props.initialUser,
    };
  }

  public componentDidMount(): void {
    this.props.client.start().then((user) => {
      this.setState({
        user,
      });
      this.props.client.on('user', this.onUser);
    });
  }

  public componentWillUnmount(): void {
    this.props.client.unsubscribe('user', this.onUser);
  }

  public render(): JSX.Element | null {
    if (!this.props.children) {
      return null;
    }
    return React.Children.only(this.props.children);
  }

  public getChildContext(): any {
    const { client: oothClient } = this.props;
    const { user } = this.state;
    return {
      oothClient,
      user,
    };
  }

  private onUser = async (user: User) => {
    this.setState({
      user,
    });
  };
}

export const withOoth = getContext({
  oothClient: PropTypes.object.isRequired,
});

export const withUser = getContext({
  user: PropTypes.object,
});
