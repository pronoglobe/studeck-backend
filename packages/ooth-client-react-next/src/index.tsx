import * as React from "react";
import { composeInitialProps } from "compose-next";
import { OothClient, User } from "ooth-client";
import { OothProvider } from "ooth-client-react";

export default (getOothClient: OothClient | (() => OothClient)) => {
  let commonOothClient: OothClient;

  if (typeof getOothClient !== "function") {
    console.warn(
      "SECURITY WARNING: Don't pass OothClient directly to ooth-client-react-next. Instad pass a function that returns a new object. This is needed for SSR."
    );
  }

  const getActualOothClient = () => {
    if (typeof getOothClient !== "function") {
      return getOothClient;
    }

    if ((process as any).browser) {
      if (!commonOothClient) {
        commonOothClient = getOothClient();
      }

      return commonOothClient;
    }

    return getOothClient();
  };

  class OothProviderWithInitialProps extends React.Component<{ initialUser: User | undefined; children: React.Component }> {
    private oothClient: OothClient;

    constructor(props: any) {
      super(props);
      this.oothClient = getActualOothClient();
    }

    public static async getInitialProps(ctx: any): Promise<{ initialUser: User | undefined }> {
      const oothClient = getActualOothClient();

      return {
        initialUser: ctx.req
          ? await oothClient.fetchUser(ctx.req.headers && { cookie: ctx.req.headers.cookie })
          : await oothClient.start()
      };
    }

    public render(): JSX.Element {
      const { initialUser, children } = this.props;

      return (
        <OothProvider client={this.oothClient} initialUser={initialUser}>
          {children}
        </OothProvider>
      );
    }
  }

  return composeInitialProps(OothProviderWithInitialProps);
};
