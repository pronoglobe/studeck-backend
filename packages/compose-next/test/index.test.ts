import { Component } from 'react';
import { composeInitialProps } from '../src';

class A extends Component {
  static getInitialProps() {
    return Promise.resolve({
      a: 1,
    });
  }
}

class B extends Component {
  static getInitialProps() {
    return Promise.resolve({
      b: 2,
    });
  }
}

describe('composeInitialProps', () => {
  test('should merge initial props', async () => {
    const Composed = composeInitialProps(A)(B);

    expect(await Composed.getInitialProps()).toMatchSnapshot();
  });
});
