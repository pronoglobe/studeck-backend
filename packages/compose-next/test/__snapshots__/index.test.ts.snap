// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`composeInitialProps should merge initial props 1`] = `
Object {
  "childProps": Object {
    "b": 2,
  },
  "parentProps": Object {
    "a": 1,
  },
}
`;
